man/ash_query.1
.\"
.\"Copyright 2011 Carl Anderson
.\"
.\"Licensed under the Apache License, Version 2.0 (the "License");
.\"you may not use this file except in compliance with the License.
.\"You may obtain a copy of the License at
.\"
.\"    http://www.apache.org/licenses/LICENSE-2.0
.\"
.\"Unless required by applicable law or agreed to in writing, software
.\"distributed under the License is distributed on an "AS IS" BASIS,
.\"WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
.\"See the License for the specific language governing permissions and
.\"limitations under the License.
.\"

.TH ash_query 1 \
  "Updated: __DATE__" \
  "__VERSION__" \
  "Advanced Shell History"


.SH NAME
ash_query - The advanced shell history query manager.


.SH SYNOPSIS
Usage: ash_query [options]
      --help
  -d  --database VALUE
  -f  --format VALUE
  -p  --print_query VALUE
  -q  --query VALUE
  -F  --list_formats
  -H  --hide_headings
  -Q  --list_queries
      --version


.SH DESCRIPTION
.B ash_query
is a part of the Advanced Shell History package.  It provides a
convenient way to execute saved queries against a command history database.

To collect command history data and to populate a history database, you must
first source the appropriate shell code into your shell session.

If you intend to use this system to manage your own personal command history,
it is recommended that you add the appropriate source command to your shell rc
file so it is automatically started with each new session.

.B For bash, add this to your ~/.bashrc file:
.RS
source /usr/lib/advanced_shell_history/bash
.RE

.B For zsh, add this to your ~/.zshrc file:
.RS
source /usr/lib/advanced_shell_history/zsh
.RE


.SH OPTIONS
.IP "      --help"

Display help and exit 0.

.IP "  -d  --database VALUE"

The filename (VALUE) of the database to query.
Typically this will be ~/.ash/history.db.
If this is not specified on the command line, ash_query will look for a shell
environment variable ASH_CFG_HISTORY_DB and try to use that.

.IP "  -f  --format VALUE"

Select an output format (VALUE) from:

.B aligned
  Columns are aligned and separated with spaces.      

.B csv
  Columns are comma separated with strings quoted.    

.B group
  Repeated values are summarized before the rows.     

.B null
  Columns are null separated with strings quoted. 

If format is not specified, ash_query will look for a default format
environment variable ASH_CFG_DEFAULT_FORMAT and try to use that.
If neither are specified, the default is 'aligned'.


.IP "  -p  --print_query VALUE"

Print the named query (VALUE) to stdout.
If the query uses shell variables, the generic query will be printed in
addition to the query after variable substitution.

.IP "  -q  --query VALUE"

Execute the named query (VALUE) and display results formatted according to
the specified --format.  If no query is specified, ash_query will check for
a named query in the shell environment variable ASH_CFG_DEFAULT_QUERY and
attempt to use that.

.IP "  -F  --list_formats"

List all the available output formats.

.IP "  -H  --hide_headings"

Suppress the headings of output tables (sometimes useful for scripting).

.IP "  -Q  --list_queries"

List the names and descriptions of all available saved queries taken from
/etc/ash/queries and ~/.ash/queries.

.IP "      --version"

Display the version number and exit.


.SH FILES
.I /etc/ash/ash.conf
.RS
Contains environment variables to be sourced into your shell.
.RE

.I /etc/ash/queries
.RS
Contains saved queries which can be invoked using the
.BR ash_query(1)
command.  Also see
.I ~/.ash/queries
for user-written queries.
.RE

.I ~/.ash/history.db
.RS
The default location for the sqlite3 database holding command history.  See
http://code.google.com/p/advanced-shell-history/w for more details on how
the data is stored internally and other tips for querying the data.
.RE

.I /usr/lib/advanced_shell_history/bash
.RS
Sourced for bash sessions.
.RE

.I /usr/lib/advanced_shell_history/zsh
.RS
Sourced for zsh sessions.
.RE


.SH ENVIRONMENT
.IP ASH_CFG_DB_FAIL_RANDOM_TIMEOUT
After a failed select, sleep a random number of milliseconds before retrying.
This is intended to add some noise to the retry mechanism.

.IP ASH_CFG_DB_FAIL_TIMEOUT
After a failed select, sleep this many milliseconds before retrying.

.IP ASH_CFG_DB_MAX_RETRIES
Quit db retries after this many failed attempts.

.IP ASH_CFG_DEFAULT_FORMAT
The default format to display queried data returned by ash_query.  Set this
to something other than 'aligned' if you prefer a different format.

.IP ASH_CFG_DEFAULT_QUERY
The default query to execute by ash_query.  Set this to the name of your
favorite query if you don't want to specify the same query name each time.

.IP ASH_CFG_HIDE_USAGE_FOR_NO_ARGS
Normally, if you invoke ash_query with no arguments, the --help output is
displayed.  With this set to a non-empty value, the --help output is
suppressed in this case.

.IP ASH_CFG_HISTORY_DB
The default database to query.  This is set by sourcing one of the shell
scripts in /usr/lib/advanced_shell_history and signifies the location
of the database where commands are logged.  If this variable exists, the
--database flag does not need to be used.

.IP ASH_CFG_IGNORE_UNKNOWN_FLAGS
Normally ash_query complains when it sees unknown flags.  With this variable
set to a non-empty value, unknown flags are ignored.

.IP ASH_CFG_LOG_DATE_FMT
If logging is in use, this format string can be set to customize the date
string.

.IP ASH_CFG_LOG_FILE
The file destination of logged messages, if logging is in use.

.IP ASH_CFG_LOG_LEVEL
The lowest level of logging to make visible.  Levels (in increasing order)
are DEBUG, INFO, WARN, ERROR and FATAL.


.SH "SEE ALSO"
.BR _ash_log(1)
for logging history


.SH AUTHOR
Carl Anderson, Google Inc.


.SH BUGS
Report bugs at http://code.google.com/p/advanced-shell-history/issues


src/ash_query.cpp
/*
   Copyright 2011 Carl Anderson

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
*/

/**
 * This program is designed to invoke saved queries in /etc/ash/queries and
 * ~/.ash/queries allowing multiple output formatting styles.
 */

#include "ash_query.hpp"

#include "command.hpp"
#include "config.hpp"
#include "database.hpp"
#include "flags.hpp"
#include "formatter.hpp"
#include "logger.hpp"
#include "queries.hpp"
#include "session.hpp"

#include <algorithm>
#include <iomanip>
#include <iostream>

using namespace ash;
using namespace flag;
using namespace std;


// TODO(cpa): add some formatting options (\0 delim, tab delim, etc).

DEFINE_string(database, 'd', 0, "A history database to query.");
DEFINE_string(format, 'f', 0, "A format to display results.");
DEFINE_string(print_query, 'p', 0, "Print the query SQL.");
DEFINE_string(query, 'q', 0, "The name of the saved query to execute.");

DEFINE_flag(list_formats, 'F', "Display all available formats.");
DEFINE_flag(hide_headings, 'H', "Hide column headings from query results.");
DEFINE_flag(list_queries, 'Q', "Display all saved queries.");
DEFINE_flag(version, 0, "Show the version and exit.");


typedef map<string, string> RowsType;


/**
 * Prints a mapping of string to string using a fixed width between columns.
 */
void display(ostream & out, const RowsType & rows, const string & name) {
  RowsType::const_iterator i, e;
  const size_t XX = 4;  // The number of spaces between columns.
  size_t widths[2] = {name.size() + XX, string("Description").size() + XX};

  // Calculate the required widths for each column.
  for (i = rows.begin(), e = rows.end(); i != e; ++i) {
    widths[0] = max(widths[0], XX + (i -> first).size());
    widths[1] = max(widths[1], XX + (i -> second).size());
  }

  // Output the headings and the rows.
  cout << left
       << setw(widths[0]) << name
       << setw(widths[1]) << "Description" << endl;
  for (i = rows.begin(), e = rows.end(); i != e; ++i) {
    cout << left
         << setw(widths[0]) << i -> first
         << setw(widths[1]) << i -> second << endl;
  }
}


/**
 * Executes a query, printing the results to stdout according to the
 * user-chosen output format.
 */
int execute(const string & sql) {
  Config & config = Config::instance();

  // Get the filename backing the database we are about to query.
  string db_file(FLAGS_database);
  if (db_file == "") {
    if (config.get_string("HISTORY_DB") == "") {
      cerr << "Expected either --database or ASH_CFG_HISTORY_DB "
           << "to be defined." << endl;
      return 1;
    }
    db_file = config.get_string("HISTORY_DB");
  }

  // Prepare the DB for reading.
  Session::register_table();
  Command::register_table();
  Database db(db_file);

  // Get the intended Formatter before executing the query.
  string format = FLAGS_format == ""
    ? config.get_string("DEFAULT_FORMAT", "aligned")
    : FLAGS_format;
  Formatter * formatter = Formatter::lookup(format);
  if (!formatter) {
    cerr << "\nUnknown format: '" << format << "'" << endl;
    display(cerr << '\n', Formatter::get_desc(), "Format");
    return 1;
  }

  // Execute the query and display any results.
  ResultSet * rs = db.exec(sql);
  formatter -> show_headings(!FLAGS_hide_headings);
  formatter -> insert(rs, cout);
  if (rs) delete rs;
  return 0;
}


/**
 * Query the history database.
 */
int main(int argc, char ** argv) {
  // Load the config from the environment.
  Config & config = Config::instance();

  if (argc == 1) {  // No flags.
    if (config.sets("DEFAULT_QUERY")) {
      return execute(config.get_string("DEFAULT_QUERY"));
    }
    if (!config.sets("HIDE_USAGE_FOR_NO_ARGS")) {
      Flag::parse(&argc, &argv, true);  // Sets the prog name in help output.
      Flag::show_help(cerr);
    }
    return 1;
  }

  // Parse the flags, removing from argv and argc.
  Flag::parse(&argc, &argv, true);

  // Abort if unrecognized flags were used on the command line.
  if (argc != 0 && !config.sets("IGNORE_UNKNOWN_FLAGS")) {
    cerr << "unrecognized flag: " << argv[0] << endl;
    Flag::show_help(cerr);
    return 1;
  }

  // Display version, if that's all that was wanted.
  if (FLAGS_version) {
    cout << ASH_VERSION << endl;
    return 0;
  }

  // Display available query names, if requested.
  if (FLAGS_list_queries) {
    display(cout, Queries::get_desc(), "Query");
    return 0;
  }

  // Initialize the available formatters.
  CsvFormatter::init();
  NullFormatter::init();
  SpacedFormatter::init();
  GroupedFormatter::init();

  // Diaplay the available format names.
  if (FLAGS_list_formats) {
    display(cout, Formatter::get_desc(), "Format");
    return 0;
  }

  // Print the requested query (both generic and actual, if different).
  if (FLAGS_print_query != "") {
    string sql = Queries::get_sql(FLAGS_print_query);
    string raw = Queries::get_raw_sql(FLAGS_print_query);
    if (raw == "") {
      cout << "Query not found: " << FLAGS_print_query << "\nAvailable:\n";
      display(cout, Queries::get_desc(), "Query");
      return 1;
    }
    cout << "Query: " << FLAGS_print_query << endl;
    if (raw != sql) {
      cout << "Template Form:\n" << raw << "\nActual SQL:\n";
    }
    cout << sql << endl;
    return 0;
  }

  // Make sure the requested query exists.
  string sql = Queries::get_sql(FLAGS_query);
  if (sql == "") {
    cout << "Query not found: " << FLAGS_query << "\nAvailable:\n";
    display(cout, Queries::get_desc(), "Query");
    return 1;
  }

  // Execute the requested query.
  return execute(sql);
}
